### PR TITLE
Support nested query for expanding path, fixes #1982

### LINKED ISSUE #1982: Routing problems
I have upgraded my new project to 0.13 and I have routing problems.

It is possible now to pass multiple optional parameters with proper nesting?
In padrino 0.12.5 I could write:

``` ruby
q = { per_page: 2, page: 4 }
url_for(:notes, :index, query: q, beta: 50)
```

which build route to:
``` url
/notes?query[per_page]=2&query[page]=4&beta=50
```

In 0.13 this is not possible anymore, it doesn't build properly:
``` url
/notes?query={:per_page=%3E%222%22,%20:page=%3E%224%22}&beta=50 
```

In new version query parameter is string in controller and it seems to not transform passed hash.
This is a bug? Is there workaround?

### PR DESCRIPTION
**Pull request by namusyaka**
`Mustermann::Expander` hasn't support nested query yet. So I've implemented it as a part of padrino. 

/cc @akemrir @padrino/core-members

### CODE REVIEW

File: padrino-core/lib/padrino-core/path_router/matcher.rb [hunk 1]
```diff
@@ -38,9 +38,35 @@ def to_regexp
...
+        when Array
+          value.map { |v|
+            build_nested_query(v, "#{prefix}[]")
+          }.join("&")
+        when Hash
+          value.map { |k, v|
+            build_nested_query(v, prefix ? "#{prefix}[#{k}]" : k)
+          }.reject(&:empty?).join('&')
+        when nil
+          prefix
```
Comment (ujifgc) [inline padrino-core/lib/padrino-core/path_router/matcher.rb]: Should be `prefix.to_s`, no? If `prefix` is nil it returns nil which does not look like a valid query.

**Comment by ujifgc** [comment]
http://www.ietf.org/rfc/rfc2732.txt states that square brackets are special and shall not be used in URI outside of IPv6 address. Maybe we should look at https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/object/to_query.rb and write it without touching core classes.

**Comment by namusyaka** [comment]
right.

**Comment by ujifgc** [comment]
Please look at c42ffc46666cf97336b5903e766be5022444fd21

**Comment by namusyaka** [comment]
@ujifgc Awsome. I've just pushed changed using your method. Could you confirm this?

**Pull request merged**